### PR TITLE
#2679 Update preferences after setting new value

### LINKED ISSUE #2679: changes to workspace preferences don't always take effect
Changes made to workspace preferences take effect when users edit the `settings.json` manually in editor.

Changes don't take effect when they are made through the pref tree.

peek 2018-08-27 12-55

### PR DESCRIPTION
**Pull request by vinokurig**
Fixes #2679

### CODE REVIEW

File: packages/preferences/src/browser/abstract-resource-preference-provider.ts [hunk 1]
```diff
@@ -72,7 +72,7 @@ export abstract class AbstractResourcePreferenceProvider extends PreferenceProvi
             const result = jsoncparser.applyEdits(content, edits);
 
             await resource.saveContents(result);
-            this.onDidPreferencesChangedEmitter.fire(undefined);
+            await this.readPreferences();
```
Comment (akosyakov) [inline packages/preferences/src/browser/abstract-resource-preference-provider.ts]: We are following the unidirectional data flow, it is updated already via: https://github.com/theia-ide/theia/pull/2707/files#diff-46af11dd5ed879ba4d08304687785b37R56

Comment (vinokurig) [reply packages/preferences/src/browser/abstract-resource-preference-provider.ts]: Done, it works!

**Comment by akosyakov** [comment]
@elaihau could you verify that it fixes the issue, please?

**Comment by akosyakov** [comment]
@elaihau thanks, let's tackle next bug with another PR

**Comment by vinokurig** [comment]
@akosyakov fixed tests

File: packages/preferences/src/browser/abstract-resource-preference-provider.ts [hunk 2]
```diff
@@ -72,14 +72,18 @@ export abstract class AbstractResourcePreferenceProvider extends PreferenceProvi
...
         }
     }
 
     protected async readPreferences(): Promise<void> {
         const newContent = await this.readContents();
         const strippedContent = jsoncparser.stripComments(newContent);
-        this.preferences = jsoncparser.parse(strippedContent);
+        const preferences = jsoncparser.parse(strippedContent);
+        if (preferences) {
+            this.preferences = preferences;
```
[the remark this answers is not in the document]
Comment (vinokurig) [reply packages/preferences/src/browser/abstract-resource-preference-provider.ts]: Agree, I'll change that
>When does it happen btw?

If `strippedContent` is empty or is a simple text (not closed in `{}`)

[the remark this answers is not in the document]
Comment (vinokurig) [reply packages/preferences/src/browser/abstract-resource-preference-provider.ts]: Done

**Pull request merged**